### PR TITLE
fix(scope-manager): incorrect reference for this within a jsx identifier

### LINKED ISSUE #4055: Reference incorrectly created for `this` within a JSXIdentifier
**Repro**

```TS
import React from 'react';

class Hello extends React.Component<{tag: () => React$MixedElement}> {
  inline() {
    return <this.props.tag />
  }
}
```

**Expected Result**

The `this` should be treated like a `ThisExpression`.

**Actual Result**

The `this` is treated like a variable called `this`, and thus creates an unclosed reference to a variable that can never exist.

**Additional Info**

Context: https://github.com/typescript-eslint/typescript-eslint/issues/3788#issuecomment-951081315
cc @ljharb

### PR DESCRIPTION
**Pull request by juank1809**
## PR Checklist

-   [x] Addresses an existing issue: fixes #4055
-   [x] That issue was marked as [accepting prs](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aopen+is%3Aissue+label%3A%22accepting+prs%22)
-   [x] Steps in [CONTRIBUTING.md](https://github.com/typescript-eslint/typescript-eslint/blob/main/CONTRIBUTING.md) were taken

## Overview

Currently when a JSX Identifier has name="this" is treated like a variable. This PR will ignore the JSXIdentifier when its name is "this" and thus not referencing it

### CODE REVIEW

**Review by bradzacher — CHANGES_REQUESTED** [review]
thanks for putting this fix up!

This is *almost* correct!
If we inspect the output from the TS and babel JSX transforms, we can see that another case we need to handle is when `this` is by itself:
```ts
import React from 'react';

class Hello extends React.Component<{tag: () => JSX.Element}> {
  inline() {
    return [
        <this.props.tag />,
        <this />,
    ];
  }
}
```
[ts playground](https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBAJQKYEMDG8BmUIjgcilQ3wG4AoctAGxQGc64AJJa6iOJADxiQDsAJo2ToYAOgDCuSH34wAPAG8YKAOYAuOAAoAlHAC8APjgApAMoANMQFFqSEHIC+xxeThxgfap6S64r93ciGABXKD44AG03QNj5GAALYDoxMBwwFJVVOAB6QwAaGNj3eKTGPMLYgF0Kd0dyRyA)
[babel playground](https://babeljs.io/repl#?browsers=defaults%2C%20not%20ie%2011%2C%20not%20ie_mob%2011&build=&builtIns=false&corejs=3.6&spec=false&loose=false&code_lz=JYWwDg9gTgLgBAJQKYEMDG8BmUIjgcilQ3wG4AoctAGxQGc64AJJa6iOJADxiQDsAJo2ToYAOgDCuSH34wAPAG8YKAOYAuOAAoAlHAC8APjgApAMoANMQFFqSEHIC-xxeThxgfap6S64r93ciGABXKD44AG03QNj5GAALYDoxMBwwFJVVOAB6QwAaGNj3eKTGPMLYgF0Kd0dyRyA&debug=false&forceAllTransforms=false&shippedProposals=false&circleciRepo=&evaluate=false&fileSize=false&timeTravel=false&sourceType=module&lineWrap=true&presets=env%2Creact%2Cstage-2%2Ctypescript&prettier=false&targets=&version=7.17.2&externalPlugins=&assumptions=%7B%7D)

So we need to handle the more generic case of when the `JSXIdentifier` has the `.name` of `"this"`

**Review by JoshuaKGoldberg — CHANGES_REQUESTED** [review]
Codecov is right - the change seems great at first glance but we can really review once there are tests please 🤗 

Let me know if it's not clear how to add them and I can write something up.

**Comment by bradzacher** [comment]
I should probably add this to the docs.
You can add tests for this by adding some `.tsx` files in this folder:
https://github.com/typescript-eslint/typescript-eslint/tree/main/packages/scope-manager/tests/fixtures/jsx

And then you can run the test using
```
$ cd packages/scope-manager
$ yarn jest fixtures.test -u
```
This will generate a corresponding `.shot` file for each `.tsx` file you add - the contents of which is the scope tree!

**Review by bradzacher — CHANGES_REQUESTED** [review]
snapshot is looking good - no dangling references to `this` in the analysis which is perfect!

---

I *think* you can simplify this a little by moving this logic into the `JSXIdentifier` selector and have it in one place.

`JSXIdentifier` is only valid in the following positions:
- `JSXOpeningElement.name` (which is analysed)
- `JSXClosingElement.name` (not analysed at all)
- `JSXMemberExpression.property` (not analysed at all)
- `JSXMemberExpression.object` (analysed only in a `JSXOpeningElement.name`)
- `JSXNamespacedName.name` (not analysed at all)
- `JSXNamespacedName.namespace` (analysed only in a `JSXOpeningElement.name`)

So by my reckoning this should be safe to just do within `JSXIdentifier`!

Otherwise this LGTM

File: packages/scope-manager/src/referencer/Referencer.ts [hunk 1]
```diff
@@ -521,10 +521,15 @@ class Referencer extends Visitor {
   protected JSXOpeningElement(node: TSESTree.JSXOpeningElement): void {
     this.referenceJsxPragma();
     if (node.name.type === AST_NODE_TYPES.JSXIdentifier) {
-      if (node.name.name[0].toUpperCase() === node.name.name[0]) {
+      if (
+        node.name.name[0].toUpperCase() === node.name.name[0] ||
+        node.name.name === 'this'
+      ) {
```
Comment (juank1809) [inline packages/scope-manager/src/referencer/Referencer.ts:527]: ``

Comment (juank1809) [inline packages/scope-manager/src/referencer/Referencer.ts:527]: But in order to be analysed the logic needs to be outside the JSXIdentifier. For example if we move the logic from JSXOpeningElement, JSXIdentifier would never be visited/analysed. 

Besides, the JSXIdentifier is agnostic whether it is being analyzed from a JSXMemberExpression or a JSXOpeningElement, so I cannot access these information inside the selector making it difficult to manage the logic inside the selector.

Sorry if I didn't understand your proposal correctly.

**Review by bradzacher — APPROVED** [review]
LGTM!

**Pull request merged**